File: unity-android/app/src/main/java/com/snap/camerakitdemos/unity/MainUnityActivity.java
package com.snap.camerakitdemos.unity;

import android.content.Intent;
import android.os.Bundle;
import android.os.Process;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;

import com.snap.samples.OverrideUnityActivity;

public class MainUnityActivity extends OverrideUnityActivity {
    // Setup activity layout
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Intent intent = getIntent();
        handleIntent(intent);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleIntent(intent);
        setIntent(intent);
    }

    void handleIntent(Intent intent) {
        if(intent == null || intent.getExtras() == null) return;

        if(intent.getExtras().containsKey("doQuit"))
            if(mUnityPlayer != null) {
                finish();
            }
    }

    @Override
    protected void invokeCameraKit(int alienHitCount) {
        Log.d("camkit", "Invoking Camera Kit " + alienHitCount);
    }

    protected void showMainActivity() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT | Intent.FLAG_ACTIVITY_SINGLE_TOP);
//        intent.putExtra("setColor", setToColor);
        startActivity(intent);
    }

    @Override public void onUnityPlayerUnloaded() {
        showMainActivity();
    }
}

File: unity-android/app/src/main/java/com/snap/camerakitsamples/unity/android/MainUnityActivity.kt
package com.snap.camerakitsamples.unity.android

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.core.graphics.component1
import com.snap.camerakit.support.app.CameraActivity
import com.snap.samples.OverrideUnityActivity
import com.unity3d.player.UnityPlayer

private const val REQUEST_CODE_CAMERA_KIT_CAPTURE = 1
private const val REQUEST_CODE_CAMERA_KIT_PLAY = 2
private const val TAG = "MainUnityActivity"

class MainUnityActivity : OverrideUnityActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = intent
        handleIntent(intent)
//        setInstance(this)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
        setIntent(intent)
    }

    private fun handleIntent(intent: Intent?) {
        if (intent == null || intent.extras == null) return
        if (intent.extras!!.containsKey("doQuit")) if (mUnityPlayer != null) {
            finish()
        }
    }

    override fun invokeCameraKitWithGroup(
        lensGroupIds: Array<String>,
        startingLensId: String?,
        cameraKitMode: Int,
        remoteApiSpecId: String?
        ) {
        val cameraKitConfig = CameraActivity.Configuration.WithLenses(
            lensGroupIds,
            startingLensId
        )
        invokeCameraKit(cameraKitMode, remoteApiSpecId, cameraKitConfig)
    }

    private fun invokeCameraKit(
        cameraKitMode: Int,
        remoteApiSpecId: String?,
        cameraKitConfig: CameraActivity.Configuration
    ) {
        val (requestCode) = if (cameraKitMode == 0) {
            REQUEST_CODE_CAMERA_KIT_PLAY
        } else {
            REQUEST_CODE_CAMERA_KIT_CAPTURE
        }
        remoteApiSpecId?.let { UnityGenericApiService.Factory.supportedApiSpecIds = setOf(it) }
        val intent = CustomCameraActivity.intentFor(this, cameraKitConfig, cameraKitMode);
        startActivityForResult(intent, requestCode)
    }

    override fun invokeCameraKitWithSingleLens(
        lensId: String,
        groupId: String,
        lensLaunchDataKeys: Array<String>?,
        lensLaunchDataValues: Array<String>?,
        cameraKitMode: Int,
        remoteApiSpecId: String?
    ) {
        val cameraKitConfig = CameraActivity.Configuration.WithLens(
            lensId,
            groupId,
            true,
            {
                if (lensLaunchDataKeys != null && lensLaunchDataValues != null) {
                    for (i in lensLaunchDataKeys.indices) {
                        this.putString(lensLaunchDataKeys[i], lensLaunchDataValues[i]);
                    }
                }
            }
        )
        invokeCameraKit(cameraKitMode, remoteApiSpecId, cameraKitConfig);
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_CODE_CAMERA_KIT_PLAY -> {
                val result = CameraActivity.Play.parseResult(resultCode, data)
                Log.d(TAG, "Got Camera Kit play result: $result")
            }
            REQUEST_CODE_CAMERA_KIT_CAPTURE -> {
                val result = CameraActivity.Capture.parseResult(resultCode, data)
                Log.d(TAG, "Got Camera Kit capture result: $result")
            }
            else -> {
                Log.d(TAG, "Got unexpected requestCode [$requestCode] " +
                        "resultCode [$resultCode] back with data: $data")
                super.onActivityResult(requestCode, resultCode, data)
            }
        }
    }
}